src/tasks/build/buildTaskProvider.ts

```diff
@@ -0,0 +1,268 @@
+
+    onDidWrite: Event<string> = this.writeEmitter.event;
+    onDidClose: Event<number> = this.closeEmitter.event;
+
+    async open(): Promise<void> {
+        // TODO: consider change to terminal name via changeNameEmitter.
+        // see: https://github.com/microsoft/vscode/issues/154146
+
+        if (this.definition.paths.length === 1 &&
+                this.definition.paths[0] === BuildTaskProvider.workspace) {
+            await this.buildWorkspace();
+        } else {
+            await this.buildProjects();
+        }
+        this.writeEmitter.fire('Task complete.\r\n');
+        this.closeEmitter.fire(0);
+    }
+
+    close(): void {
+        this.cancellationTokenSource.cancel();
+        this.cancellationTokenSource.dispose();
+    }
+
+    async buildWorkspace(): Promise<void> {
+        this.writeEmitter.fire("Building all the Java projects in workspace...\r\n\r\n");
+        try {
+            await commands.executeCommand(Commands.COMPILE_WORKSPACE, this.definition.isFullBuild, this.cancellationTokenSource.token);
+        } catch (e) {
+            if (checkErrorsReportedByJavaExtension()) {
+                commands.executeCommand(Commands.WORKBENCH_VIEW_PROBLEMS);
+                this.writeEmitter.fire("Errors found when building the workspace.\r\n\r\n");
+            } else {
+                this.writeEmitter.fire("Errors occur when building the workspace:\r\n");
+                this.writeEmitter.fire(`${e}\r\n\r\n`);
+            }
+        }
+    }
+
+    async buildProjects(): Promise<void> {
+        // tslint:disable-next-line: prefer-const
+        let [includedPaths, excludedPaths, invalidPaths] = categorizePaths(this.definition.paths, this.scope);
+        if (invalidPaths.length) {
+            this.printList("Following paths are invalid, please provide absolute paths instead:", invalidPaths);
+            return;
+        }
+
+        const projectUris: string[] = await Jdtls.getProjectUris();
+        const projectPaths: string[] = projectUris
+            .map(uri => Uri.parse(uri).fsPath)
+            .filter(p => path.basename(p) !== "jdt.ls-java-project");
+        [includedPaths, invalidPaths] = getFinalPaths(includedPaths, excludedPaths, projectPaths);
+
+        if (invalidPaths.length) {
+            this.printList("Following paths are skipped due to not matching any project root path:", invalidPaths);
+        }
+
+        if (includedPaths.length === 0 || this.cancellationTokenSource.token.isCancellationRequested) {
+            return;
+        }
+
+        this.printList("Building following projects:", includedPaths);
+        const uris: Uri[] = includedPaths.map(p => Uri.file(p));
+        try {
+            const res = await commands.executeCommand(Commands.BUILD_PROJECT, uris, this.definition.isFullBuild,
+                this.cancellationTokenSource.token);
+            if (res === Jdtls.CompileWorkspaceStatus.Witherror && checkErrorsReportedByJavaExtension()) {
```

Comment:
I suggest to use `closeEmitter.fire()` to indicate the task result, in this way the user can easily distinguish task result in the terminal  UI. e.g, https://github.com/microsoft/vscode-gradle/blob/a9d5977d0a8505ddc5bf55000046e0e96428e4d3/extension/src/terminal/GradleRunnerTerminal.ts#L105-L108

Reply:
Do you mean call `closeEmitter.fire()` in the finally block?

Reply:
currently we always use fire(0) (see: https://github.com/microsoft/vscode-java-dependency/blob/2759b924f89347439f5b31a77130845c764cfb14/src/tasks/build/buildTaskProvider.ts#L97) and it will return a successful task state, if we fire another value, e.g., fire(1) for failed condition, the failed state can be shown.

Reply:
Got your point, good suggestion!